### PR TITLE
BaseStrategy contract

### PR DESCRIPTION
**Pull request by brkomir**
Implemented `BaseStrategy` contract to hold shared functionality needed by any strategy implementation.

### CODE REVIEW

File: contracts/strategy/BaseStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * Sets up the essential fields required for any strategy implementation.
+     *
+     * @notice This is ment to be called only from the #constructor of the implemeting contracts.
+     *
+     * @param _vault address of the vault that will use this strategy
+     * @param _underlying address of the underlying token
+     * @param _admin address of the administrator account
+     */
+    function setup(
```
Comment (gabrielpoca) [inline contracts/strategy/BaseStrategy.sol:61]: We have been starting internal functions with "_"

File: contracts/strategy/BaseStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * Transfers administrator rights for the Strategy to another account,
+     * revoking current admin roles and setting up the roles for the new admin.
+     *
+     * @notice Can only be called by the account with the ADMIN role.
+     *
+     * @param _newAdmin The new ADMIN account.
+     */
+    function transferAdminRights(address _newAdmin) external virtual;
```
Comment (gabrielpoca) [inline contracts/strategy/BaseStrategy.sol:104]: What's the reason for this function not to be implemented?

Comment (brkomir) [reply contracts/strategy/BaseStrategy.sol:104]: the reason is that strategies can have other roles associated with an admin account (ex YearnStrategy) and that roles should be also transferred to the new admin, so the idea was for strategies to provide an implementation for this and call `changeAdmin` internal function within that implementation. However, I think we can also provide a default implementation for this, as you might be suggesting here

Comment (gabrielpoca) [reply contracts/strategy/BaseStrategy.sol:104]: hummmm, I just wonder if it's worth having this common interface. You could just have the `transferAdminRights` public function and any strategy could either keep it or call it from another function right?

Comment (brkomir) [reply contracts/strategy/BaseStrategy.sol]: that's one option. My intent was to retain the same interface meaning there is only one function addressing transferring of the admin rights. Check out this [commit](https://github.com/lindy-labs/sc_solidity-contracts/pull/181/commits/a84182b281ccffa84366264a94ec2135c02f2fb9)

**Pull request merged**